### PR TITLE
Make 'LoginUtils.requireLogin()' to not block the execution.

### PR DESCRIPTION
**Pull request by vrubezhny**
Checking whether the user is logged in should not block 'oc' binary execution in case login is required

### CODE REVIEW

File: src/util/loginUtil.ts [hunk 1]
```diff
@@ -24,8 +24,11 @@ export class LoginUtil {
      * @returns true if the user needs to log in to access the cluster, and false otherwise
      */
     public async requireLogin(): Promise<boolean> {
-        return await Promise.any([Oc.Instance.getCurrentUser(), Odo.Instance.getProjects()])
-            .then(() => false)
-            .catch(() => true);
+        if (await Oc.Instance.isLoggedIn()) {
+            return await Promise.any([Oc.Instance.getCurrentUser(), Odo.Instance.getProjects()])
+                .then(() => false)
+                .catch(() => true);
```
Comment (vrubezhny) [inline src/util/loginUtil.ts:30]: `Oc.Instance.isLoggedIn()` actually does the same as `Oc.Instance.getCurrentUser()`  but in not-blocking way, so probably we don't don't need this extra method, but we have to make `Oc.Instance.getCurrentUser()`  to be non-blocking instead. 

The thing I'd like to ask here... Once `oc whoami` returns the logged in user, should we continue checking whether no error happens on `odo list project` execution?

Comment (datho7561) [reply src/util/loginUtil.ts:30]: I think we should be good to remove this function and replace it with `Oc.Instance.isLoggedIn()` everywhere it's used.

File: src/oc/ocWrapper.ts [hunk 2]
```diff
@@ -455,4 +456,20 @@ export class Oc {
...
+    /**
+     * Checks if we're logged in. If not, any other async call to `oc`/`odo` binary might be blocked on
+     * waiting for user input
+     *
+     * @return 'true' if a user is logged in to a culster, 'false' otherwise
+     */
+    async isLoggedIn(): Promise<boolean> {
+        try {
+            const result: string = await CliChannel.getInstance().executeSyncTool(
+                new CommandText('oc', 'whoami'), { tomeout: 500 }  as CommonOptions);
```
Comment (datho7561) [inline src/oc/ocWrapper.ts:469]: Please fix or remove the timeout

**Review by datho7561 — CHANGES_REQUESTED** [review]
I did some testing on Windows and non-OpenShift clusters, and it seems to be working there as well (with the caveat that you can't log in/out of non-OpenShift clusters).

**Comment by vrubezhny** [comment]
> I think the reason this patch work is that it uses `execFile` instead of `exec`. `exec` creates a shell to run the program in, whereas `execFile` directly executes the executable.  The `oc` binary is likely able to detect whether or not it's being run in a shell, so it won't produce the prompt for username and password when not being invoked through a shell. This might cause issues on Windows; I'll have to double check.

May be, but `oc` is actually prompts  for a username to be entered, which you can check  in debugger. The only thing changed here is a timeout added, which kills the `oc` process when timeout expires - thus resolving the hang-up problem 

> 
> I don't think this timeout does anything, since it's mispelt: https://github.com/redhat-developer/vscode-openshift-tools/pull/3674/files#diff-fb06bee2638b8012f20db654d2191ea1dd8448a73ebe4da36750b489a188ecbbR469 .

There is 10 seconds default value for the timeout used... So, misspelling actually doesn't break the things (But yes, it is misspelled)

**Comment by datho7561** [comment]
> I thought we can make oc.Instance.getCurrentUser() to use execFile+timeout so it can be used to check if a user is logged in in LoginUtil.Instance.requireLogin() without any additional calls to Odo.Instance.getProjects() and adding any new isLoggedIn() method...
Good call. I like that refactoring better.

Maybe this will end up fixing the issue Mohit has been running into (and me occasionally) where the view takes a while to load if the kubecontext is set to an inaccessible cluster.

**Comment by vrubezhny** [comment]
> Maybe this will end up fixing the issue Mohit has been running into (and me occasionally) where the view takes a while to load if the kubecontext is set to an inaccessible cluster.

I  don't think so... I haven't dig into this yet, maybe it'll speed up the things a little bit, but not too much.

**Comment by vrubezhny** [comment]
> where the view takes a while to load if the kubecontext is set to an inaccessible cluster.

If a cluster "is inaccessible" we're actually aren't able to speed up anything. We still need to log in and this will take as much time as it takes (if possible at all).  The only thing we can do is limiting the time to connect/login with a timeout (and make such timeout to be configurable through a user preference - because the normal  time needed to login to different clusters  may vary a lot, so it's difficult to imagine/calculate what default value could it be).

**Comment by vrubezhny** [comment]
> I did some testing ...(with the caveat that **_you can't log in/out of non-OpenShift clusters_**).

Do we have any special binary/API to be invoked for non-OpenShift clusters? I might be missing something but I thought we always use `oc login` / `oc logout`, don't we?

**Comment by datho7561** [comment]
Most non-OpenShift clusters don't have a username/password system. Instead, you configure an access token in your kubeconfig, and any Kubernetes clients (eg. `kubectl`, `oc`, any libraries) should use those.

**Comment by vrubezhny** [comment]
> Most non-OpenShift clusters don't have a username/password system. Instead, you configure an access token in your kubeconfig, and any Kubernetes clients (eg. `kubectl`, `oc`, any libraries) should use those.

So, the solution might be the following:
- call  `oc logout`, then:
   - if this succeeds: leave as is (if it was an openshift-like cluster,  the  :"current-context" property will not be modified on `~/.kube/config`)
   - if this fails: this is highly-likely a non-openshift cluster, so execute `oc config  unset current-context` (this will clear the  "current-context" property, making OS Tools to show "Login" and other buttons.

**Comment by vrubezhny** [comment]
The demo of login/logout to Kind and OS Clusters, and switching to the these clusters in case we're already logged in:
[Screencast from 2023-12-15 20-04-45.webm](https://github.com/redhat-developer/vscode-openshift-tools/assets/620781/76e0e5c3-7f0e-4723-af31-58a799b17869)

**Comment by vrubezhny** [comment]
@datho7561 Could you, please, take a look again. Now 'Login' works (during my tests) for any type of cluster including Kind and Sandbox, the only problem is a ~3 seconds delay that may occur while checking whether we're logged in in case we're logged in to an OS cluster (potentially any remote/slow/etc. cluster) where `oc status` takes about 3 seconds to successfully complete in my case.

File: src/cli.ts [hunk 3]
```diff
@@ -70,4 +70,33 @@ export class CliChannel {
...
+            options.maxBuffer = 2 * 1024 * 1024;
+        }
+        if (options.timeout === undefined) {
+            options.timeout = 10 * 1000; // 10 seconds
+        }
+        if (options.encoding === undefined) {
+            options.encoding = 'utf-8';
+        }
+        if (options.killSignal === undefined) {
+            options. killSignal = 'SIGILL'
```
Comment (datho7561) [inline src/cli.ts:95]: `SIGILL` -> `SIGKILL`

File: test/integration/ocWrapper.test.ts [hunk 4]
```diff
@@ -221,27 +231,35 @@ suite('./oc/ocWrapper.ts', function () {
...
+            if (isOpenShift) {
+                await Oc.Instance.loginWithUsernamePassword(
+                    clusterUrl,
+                    username,
+                    password,
+                );
+                const currentUser = await getCurrentUser();
+                expect(currentUser.trim()).to.equal(username);
+            } else {
+                this.skip
```
Comment (datho7561) [inline test/integration/ocWrapper.test.ts:251]: I think you need to invoke `skip` as a method in order for it to work

**Review by datho7561 — CHANGES_REQUESTED** [review]
Sorry about the delay in the review. This is working very well! I found two small things in the code.

File: src/cli.ts (hunk 3 above)
Comment (vrubezhny) [reply src/cli.ts:95]: Interesting.... However this really does work... 

ACK

File: test/integration/ocWrapper.test.ts (hunk 4 above)
Comment (vrubezhny) [reply test/integration/ocWrapper.test.ts:251]: Yes, you're right

**Review by datho7561 — APPROVED** [review]
Looks good and works great. Thanks Victor!

**Pull request merged**